[WF-566] Slider component

blurbyte opened the pull request:
# New Slider component

## [WF-566](https://datacamp.atlassian.net/browse/WF-566)

## Changes

Implemented new `Slider` component. Under the hood it uses `react-range` lib (only 5KB gzipped), some APIs and internal implementations are dictated by this lib.

Thoughts:

- Inputs handled in more SOLID than DRY way (I guess it's fine for readability)
- Might be better to have `value` to be `number | [number, number]` but it's kind of hard to sync it with lib, considering onChange callback, not really sure if's worth extra effort here
- For input width calculation some are hardcoded, and ideally probably should be taken from components itself, but not really sure if it's worth spending extra time here and complicating it even further

### 🚀 Major Changes

- Added `Slider` component:
  - it's controlled component so state is handled in the parent component
  - accepts array as a `value`: when one number is passed (e.g. `[10]`) single thumb / handle is shown, but for two numbers (e.g. `[10, 30]`) it shows range between two thumbs
  - min, max, and step (fraction could be set too) are configurable
  - all labels could be hidden with `hideLabels`
  - inputs to enter values directly could be displayed with `showInputs` (not affected by `hideLabels`), they are shown instead of current value(s)
  - error handling for inputs
  - inverted, disabled, focus styles are available
  - labels could be formatted with `formatLabel`, when e.g. currency must be selected
 
### ❓ Misc Changes

- Added `react-range` dependency to handle complex slider logic
- Bumped `floating-ui` version

## Screenshot(s):

When everything is fine:

<img width="780" alt="slider" src="https://user-images.githubusercontent.com/20565536/191236309-e6021541-4b02-45a3-85e9-ee9da79e6161.png">
Some errors:

<img width="1252" alt="Screenshot 2022-09-20 at 11 40 14" src="https://user-images.githubusercontent.com/20565536/191236330-0c583667-da48-448a-9fe2-4cbad0122371.png">
<img width="1242" alt="Screenshot 2022-09-20 at 11 39 50" src="https://user-images.githubusercontent.com/20565536/191236332-e488d9e4-9687-4022-852c-ec2824de0f7d.png">

ixTec reviewed (changes requested):
Honestly looks great to the eyes, such a clean and easy user interaction!

Added a couple of comments / thoughts

- Inputs should likely just have a `min-width` instead of `width`, because there's some overflow issues with large number lengths 
    - Content should also probably be centre aligned?
    - ![Screenshot 2022-09-02 at 14 29 30](https://user-images.githubusercontent.com/9938773/188156474-3bc28c10-141b-4280-b794-ffaf205a000d.png)
- I am _really_ not a fan of the term `Thumb`, and although I can see it's the name used by `react-range`, I would choose an alternative for our prop/naming for clarity, perhaps `Handle` or `Slider` itself, with the actual waffles component being renamed to `Interactive Range` or `Range Slider`. Something to discuss during work week perhaps

blurbyte commented:
Agree with `Thumb` - personally not a big fan of it too, but decided to it because `react-range` using it, and Material Design guidelines probably popularized this term. Will change it to `Handle`.
For inputs I think their widths will have to be calculated up front based on max / step values. It's super awkward if inputs change their size when typing (this would happen with `min-width`. Since values are numbers only it should be rather easy to implement (just digits count multiplied by some hardcoded value). I guess we shouldn't rally care about incorrect values here.

blurbyte commented on src/slider/inputs.tsx:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  function renderInput(
+    value: string | number,
```
Can't do that because, e.g. case when user want to type decimal `10.` it would have to be parsed to `10` immediately so `.` would be gone making whole experience very awkward.

ixTec replied:
interesting issue - I hadn't considered this!

blurbyte commented on src/slider/slider-context.tsx:
```diff
@@ -0,0 +1,34 @@
+function useSlider() {
+  const context = useContext(SliderContext);
+
+  if (context === undefined) {
```
I guess it doesn't really hurt do have it there - could be easy to forget in future if we will decide to expose this hook.

blurbyte commented on src/slider/inputs.tsx:
```diff
@@ -0,0 +1,119 @@
+  // Clear all inputs errors when Slider is dragged
+  useEffect(() => {
+    setInputValue(value);
+    setHasError(initalError(isSingleInput));
```
Time to invest in some spell checker plugin for VSCode :)

blurbyte commented on src/slider/limit-labels.tsx:
```diff
@@ -0,0 +1,21 @@
+  formatLabel: (value: number) => number | string;
+};
+
+function LimitLabels({ formatLabel }: LimitLabelsProps) {
```
Was my initial thought for its name too, but was worried it would imply it has something to do with selecting custom range.

blurbyte commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,73 @@
+        {!hideLabels && !showInputs && (
+          <ValueLabels formatLabel={formatLabel} />
+        )}
+        <Range
```
Good idea 👍

blurbyte commented on src/slider/styles.ts:
```diff
@@ -0,0 +1,148 @@
+  return css`
+    // Shift track line slightly so thumb aligns with labels nicer
+    width: calc(100% + ${THUMB_DOT_RADIUS}px);
+    margin-left: -${THUMB_DOT_RADIUS / 2}px;
```
I prefer to reserve `calc` for calculating dynamic values, instead of something I know during design.

blurbyte commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,73 @@
+  'aria-label': ariaLabel,
+  formatLabel = (value) => value,
+}: SliderProps) {
+  if (value.length > 2 || value.length === 0) {
```
Yeah, but it's kind of hard to make it compatible with `react-range` and `onChange` handler without some ugly TS casts. So not sure if it's worth the trouble.
Also not 100% convinced if it's good idea to offer single number or array depending on type of Slider you want to have. Especially for `onChange` you will have to remember to use the proper one again.

ixTec commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,187 @@
+  /* Current slider value. Array of values. For single handle provide single value e.g. `[20]`, and for range provide two values, e.g. `[20, 50]`. */
+  value: number[];
+  /* Handler called when handle is moved. */
+  onChange: (value: number[]) => void;
```
Really not sure we want onChange and onChangeEnd still - they're used for similar purposes internally, except within the react-range component. We likely just care to expose one of the two as a prop

blurbyte replied:
I think we need both:
- onChange is used to set a value (and is constantly updating when user moves a handle)
- onChangeEnd is used to retrieve the last value after handle stops moving
So the first one is not really feasible if you want for example send an XHR request with a slider value.

blurbyte replied:
After discussion: we decided to remove this onChangeEnd callback.

ixTec reviewed (changes requested):
Looks like there's a handful of new issues and I've added some additional comments too:

- Input fields should probably show numbers to the the full decimal places, to avoid jumping between character lengths
- The error state still isn't very clear imo - perhaps we need design consultation because when an input is invalid, it gives no reason for why, including no indication for screen readers
- Story file names can probably be a bit more detailed e.g. `slider-with-inputs` instead of 'slider-input`

blurbyte commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,187 @@
+    }
+  }
+
+  function handleSecondInputChange(event: React.ChangeEvent<HTMLInputElement>) {
```
Moved them to `Input` component, should be easier to digest.

blurbyte commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,187 @@
+    setHasError(initialError(isSingleInput));
+  }, [value, isSingleInput]);
+
+  // Separate each input handler for readability
```
Added some additional comments.

blurbyte commented on src/slider/value-labels.tsx:
```diff
@@ -0,0 +1,24 @@
+
+function ValueLabels({ formatLabel }: ValueLabelsProps) {
+  const { value, inverted } = useSlider();
+  const isSingleValue = value.length === 1;
```
I would prefer to leave it at is is to be super explicit.

blurbyte commented:
A few improvements:
- added input error messages below Slider, they work nicely with screen reader
- moved each input handler from Slider to Inputs
- fixed an issue with an input when it was impossible to type value like `10.00001`

ixTec commented on src/slider/inputs.tsx:
```diff
@@ -0,0 +1,161 @@
+
+  // Separate each input handler for readability
+
+  // Short overview of how each handler works:
```
We could do with some more punctuation in these comments, just to help readability etc. See suggestion below:

``` 
  // 1. Only numeric values are allowed
  // 2. When user starts typing, any error for the Input is cleared
  // 3. For better UX, some additional checks are performed before setting final value on Slider. So sometimes Slider and Input values are out of sync, e.g. value ending with dot
  // 4. If Input value passes checks, parsed value is set on slider
  // 5. Otherwise an appropriate error is displayed
  ```

ixTec commented on src/slider/inputs.tsx:
```diff
@@ -0,0 +1,161 @@
+      onChange([parsedValue, value[1]]);
+    } else if (isNotWhitelisted(targetValue)) {
+      if (parsedValue < min || parsedValue > value[1]) {
+        setError(['First input value is out of allowed range.', error[1]]);
```
is _outside_ of

ixTec commented on src/slider/inputs.tsx:
```diff
@@ -0,0 +1,161 @@
+      if (parsedValue < min || parsedValue > value[1]) {
+        setError(['First input value is out of allowed range.', error[1]]);
+      } else {
+        setError(['First input value is not overlapping with step.', error[1]]);
```
Maybe `First input value is a valid step increment` - if I'm understanding this error correctly

ixTec commented on src/slider/inputs.tsx:
```diff
@@ -0,0 +1,161 @@
+      } else {
+        setError([
+          error[0],
+          'Second input value is not overlapping with step.',
```
Maybe `Second input value is a valid step increment` - if I'm understanding this error correctly

ixTec commented on src/slider/inputs.tsx:
```diff
@@ -0,0 +1,161 @@
+      onChange([value[0], parsedValue]);
+    } else if (isNotWhitelisted(targetValue)) {
+      if (parsedValue < value[0] || parsedValue > max) {
+        setError([error[0], 'Second input value is out of allowed range.']);
```
is outside of

ixTec commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,117 @@
+import Errors from './errors';
+
+type SliderProps = {
+  /* Current slider value. Array of values. For single handle provide single value e.g. `[20]`, and for range provide two values, e.g. `[20, 50]`. */
```
Maybe reword as "Current slider value(s) as an array. For a single-value slider, provide a single value array e.g. `[20]`, and for a range slider provide two values, e.g. `[20, 50]`".

ixTec commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,117 @@
+  value: number[];
+  /* Handler called when handle is moved. */
+  onChange: (value: number[]) => void;
+  /* The minimum allowed value of the slider. Can be decimal or negative. */
```
"Can be _a_ decimal or..."

ixTec commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,117 @@
+  /* The minimum allowed value of the slider. Can be decimal or negative. */
+  /* @default 0 */
+  min?: number;
+  /* The maximum allowed value of the slider. Can be decimal or negative. */
```
"Can be _a_ decimal or..."

ixTec commented on src/slider/slider.tsx:
```diff
@@ -0,0 +1,117 @@
+  /* The maximum allowed value of the slider. Can be decimal or negative. */
+  /* @default 100 */
+  max?: number;
+  /* The minimal distance between values. Can be decimal. */
```
"Can be _a_ decimal."

ixTec commented on src/slider/utils.ts:
```diff
@@ -0,0 +1,49 @@
+  return /^-?([0-9]{1,})?(\.)?([0-9]{1,})?$/.test(value);
+}
+
+// Check if value could be safely set on slider
```
"..._can_ be..."

ixTec reviewed (changes requested):
I would say it's almost there!

Couple of minor comments in the code and the one below, before we're good to merge:

- Regarding the error messages for our inputs, we can likely set these strings to CONST's or util literals so we can just populate whether it's first/second input. this would remove the current duplication we have

We can definitely make some further improvements later on as a separate task :)

ixTec reviewed (approved):
Let's mergeeee 🥇
